src/gluonts/ev_with_decorators/metrics.py

```diff
@@ -0,0 +1,128 @@
+# Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License").
+# You may not use this file except in compliance with the License.
+# A copy of the License is located at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# or in the "license" file accompanying this file. This file is distributed
+# on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+# express or implied. See the License for the specific language governing
+# permissions and limitations under the License.
+
+import numpy as np
+
+from gluonts.ev_with_decorators.api import Input, metric
+
+
+# POINT METRICS (two-dimensional in univariate case)
+
+
+@metric(Input("target"))
+def abs_target(target):
+    return np.abs(target)
+
+
+@metric(Input("prediction_mean"))
+def abs_prediction_mean(prediction_mean):
+    return np.abs(prediction_mean)
+
+
+@metric(Input("target"), Input("prediction_mean"))
+def error_wrt_mean(target, prediction_mean):
+    return target - prediction_mean
+
+
+@metric(Input("target"), Input("prediction_mean"))
+def error_wrt_median(target, prediction_mean):
+    return target - prediction_mean
+
+
+@metric(error_wrt_median)
+def abs_error(error_wrt_median):
+    return np.abs(error_wrt_median)
+
+
+@metric(error_wrt_mean)
+def squared_error(error_wrt_mean):
+    return error_wrt_mean * error_wrt_mean
+
+
+# LOCAL METRICS (one-dimensional in univariate case)
+
+
+@metric(squared_error)
+def mse(squared_error):
+    return np.mean(squared_error, axis=1)
+
+
+@metric(mse)
+def rmse(mse):
+    return np.sqrt(mse)
+
+
+@metric(abs_error)
+def abs_error_sum(abs_error):
+    return np.sum(abs_error, axis=1)
+
+
+@metric(abs_error)
+def abs_target_sum(abs_target):
+    return np.sum(abs_target, aixs=1)
+
+
+@metric(abs_error)
+def abs_target_mean(abs_target):
+    return np.mean(abs_target, aixs=1)
+
+
+@metric(abs_error, abs_target)
+def mape(abs_error, abs_target):
+    return np.mean(abs_error / abs_target, axis=1)
```

Comment:
So there is no distinction between a point-value and aggregations, both are `metric`?

How can I control over what axis to aggregate?

Reply:
Yes, all metrics are of the same type. The standard metrics (like MSE and MAPE) are usually considered per data entry which is why axis=1 is hardcoded here. If one would like to aggregate over all data at the same time stamp (axis=0), we could pretty much copy the metric and call it `mape_over_time` with the desired axis.

Reply:
> usually considered per data entry

Where is that assumption made?

> we could pretty much copy the metric and call it `mape_over_time` with the desired axis

Please let's not do that. 

In the example I've shared, the idea is to make axis a parameter:

```py
@aggregation(abs_error, abs_target)
def mape(abs_error, abs_target, axis):
    return np.mean(abs_error / abs_target, axis=axis)
```

The nice thing is that if I want what you call 'globa' metric, I can pass `axis = None` and it just aggregates everything to just one number. However, I also have the freedom to control if I want to aggregate by time or series. There is no need for any code duplication there.

Reply:
> Where is that assumption made?

The current `Evaluator` class evaluates metrics sequentially, one after the other. This means, every metric **has** to be aggregated per data entry first, the way things are done now. That's a restriction that should be resolved with this new approach. But still, the `Evaluator` has survived a long time and its underlying assumption "most metrics are aggregated per data entry" seems to be okay for most cases.

> In the example I've shared, the idea is to make axis a parameter:

Right, I took your initial example and modified it so that there was no variable axis left over in the end. I see how it can be useful to dynamically provide the axis but it can also be restrictive. Let's consider this example:
```
losses = evaluate([rmse], {"label": label, "prediction": prediction})
print(losses.aggregate(axis=0))
print(losses.aggregate(axis=1))
```
It lacks some flexibility in the kind of metrics we want to use. For example, we might want some sort of time slicing (using aixs=0), like summing the absolute error over each time stamp. But if that's the only metric we care about in that axis, it's wasteful to calculate all other metrics using this axis as well.
Also, I don't see how aggregating multiple times over different axis (like the mean MSE discussed above) would work if we provide the axis only at the end.

> There is no need for any code duplication there.

I'm not suggesting that we should provide a metric for every potential use case. "Mape over time" sounds very non-standard, for example. But if someone would need this kind of metric, **they** can make their own metric which just works exactly the way they want it (without having to pass an axis).

Reply:
My current philosophy which is used in the latest commit is “Everything is a metric and it can depend on anything it wants, even aggregated metrics”. Instead of that, we could also have a clear two-step process like you proposed originally:
First, we calculate all the “*point metrics*”, which are just two-dimensional NumPy arrays (in the univariate case).
Then, we calculate all *aggregated metrics* which are based only on point metrics. Each metric can be given an axis that should be used for aggregation.

I mainly see two slight issues with this:

* So, the Evaluator currently performs *two* aggregations one after the other.  First, metrics are aggregated per dataset entry, and then these values are boiled down into a single number in [get_aggregate_metrics](https://github.com/awslabs/gluonts/blob/d5bf5bf147a52798104f4c254d246ed863a68584/src/gluonts/evaluation/_base.py#L426) (https://github.com/awslabs/gluonts/blob/d5bf5bf147a52798104f4c254d246ed863a68584/src/gluonts/evaluation/_base.py#L426). These aggregations of aggregations are pretty much all the Evaluator does currently and they can actually be helpful for getting an overview of how well a model is doing on a test dataset. I think it would be nice to keep supporting something like this. (However, I could also agree on leaving further aggregation up to the user. As you said earlier, taking the mean of some values is easily done.)
* It seems like some metrics only make sense with a particular axis. The seasonal error for example doesn’t seem to make much sense if we aggregate over multiple dataset entries using the same time stamp. So, if we are interested in metrics A, B, C, D, E and we want to aggregate A and E over one axis and B, C and D over the other, this should be possible (as opposed to calculating all metrics for each axis and not using half the results). Usage like print(losses.aggregate(axis=0)) seems to invite the calculation of unused metrics.

But the latter is a minor concern. My biggest question is: Do you think aggregated metrics should be able to depend on other aggregated metrics or not?